### PR TITLE
Migrate from convention to extension

### PR DESCRIPTION
**Pull request by rougsig**
**Background**
Gradle convention does not support dsl generation. [For automatic dsl generation gradle requires extensions](https://docs.gradle.org/current/userguide/implementing_gradle_plugins.html#extensions_vs_conventions). In that reason while we use convention, we need to write dsl accessors by hand (see [ProtobufConfiguratorExts.kt](https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt)).

**Changes**
1. Migrated from convention to extension.
2. Removed `Android/JavaGenerateProtoTaskCollection`. Now it one class named `GenerateProtoTaskCollection`.
3. Removed unnecessary kotlin dsl accessors, now it generated by gradle.

**Breaking changes**
1. In reason `Changes.1`. Now `project.protobuf.protobuf` is not valid code. User should write `project.protobuf`.
2. In reason `Changes.3`. User should remove those imports from kotlin gradle scripts:
. `com.google.protobuf.gradle.protobuf`
. `com.google.protobuf.gradle.protoc`
. `com.google.protobuf.gradle.plugins`
. `com.google.protobuf.gradle.generateProtoTasks`
. `com.google.protobuf.gradle.builtins`
. `com.google.protobuf.gradle.ofSourceSet`
. `com.google.protobuf.gradle.ofFlavor`
. `com.google.protobuf.gradle.ofBuildType`
. `com.google.protobuf.gradle.ofVariant`
. `com.google.protobuf.gradle.ofNonTest`
. `com.google.protobuf.gradle.ofTest`

**Test plan**
Green pipelines. Production code was changed. Changes does not affect any plugin logic.

References issues: https://github.com/google/protobuf-gradle-plugin/issues/505

### CODE REVIEW

File: src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt [hunk 1]
```diff
@@ -24,12 +24,12 @@ import org.gradle.kotlin.dsl.invoke
...
  *
  * @receiver [Project] The project for which the plugin configuration will be applied
- * @param action A configuration lambda to apply on a receiver of type [ProtobufConfigurator]
+ * @param action A configuration lambda to apply on a receiver of type [ProtobufExtension]
  *
  * @return [Unit]
  */
-fun Project.protobuf(action: ProtobufConfigurator.()->Unit) {
-    project.convention.getPlugin(ProtobufConvention::class.java).protobuf.apply(action)
+fun Project.protobuf(action: ProtobufExtension.() -> Unit) {
```
Comment (rougsig) [inline src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt:31]: When we remove the extension registration from afterEvaluate, we can remove that extension function.

**Comment by rougsig** [comment]
**Breaking changes**
1. Removed `protobuf.protobuf` property 
User should replace `project.protobuf.protobuf` -> `project.protobuf`.
2. Removed a bunch of `com.google.protobuf.gradle.*` kotlin extensions.
User should remove unresolved imports from gradle scripts. Now that extensions autogenerated by gradle.
TODO: provide full list of removed imports
3.

**Comment by rougsig** [comment]
> Do you think we use separate extensions for Android vs non-Android? That would mean they'd have to have different names, like project.protobuf and project.protobufAndroid?

For now, I've merged them into one. As it was with kotlin. i.e. for android selectors (ofFlavour, etc.) inside a java project, will be returned an empty list of tasks

File: testProjectBase/build_base.gradle [hunk 2]
```diff
@@ -76,10 +76,18 @@ def assertFileExists(boolean exists, String path) {
 }
 
 test.doLast {
-  assert ['generateProto', 'generateGrpcProto', 'generateTestProto'] as Set ==
-      protobuf.generateProtoTasks.all().collect({ it.name }) as Set
+  assert [
```
Comment (ejona86) [inline testProjectBase/build_base.gradle:79]: Revert? I don't see any change here.

**Review by ejona86 — COMMENTED** [review]
This looks fair. Just a few comments.

**Comment by rougsig** [comment]
Any updates?

**Review by YifeiZhuang — APPROVED** [review]

**Review by ejona86 — APPROVED** [review]
Oops. I had looked over the changes after the update but had forgotten to approve.

**Pull request merged**